### PR TITLE
[v2.9][v1.29] Rancher K8s v1.29.0 update

### PR DESCRIPTION
**Pull request by krunalhinguu**
## Issue: 

- https://github.com/rancher/rancher/issues/43110

## Problem

-  To support Kubernetes 1.29 we need to bump all the k8s libraries to 1.29

## Solution

- Bumping all Kubernetes libraries to 1.29
- Removed all dependencies related to `PodSecurityPolicy`
- Removed all dependencies related to `PodSecurityPolicyTemplateProjectBinding`
- Removed all dependencies related to `PodSecurityPolicyTemplateName`
- Removed code to enable `PSP`
- `forceAdopt` is renamed to `takeOwnership` as part of helm `3.14.3`
- Bumped `golangci-lint` version. https://github.com/golangci/golangci-lint/issues/4273

 
## Testing

rke2
rke
k3s

### Automated Testing

* Test types added/modified:
    * Unit
    * Integration (Go Framework)
    * Integration (v2prov Framework)
    * Validation (Go Framework)
    * Other - Explain: _EXPLAIN_
    * None
    * _REMOVE NOT APPLICABLE BULLET POINTS ABOVE_
* If "None" - Reason: _EXPLAIN THE REASON_
  
* If "None" - GH Issue/PR: _LINK TO GH ISSUE/PR TO ADD TESTS_

Summary: _TODO_

### Regressions Considerations

_TODO_

Existing / newly added automated tests that provide evidence there are no regressions:
* _TODO_

### CODE REVIEW

File: go.mod
Comment (kinarashah) [inline go.mod]: Could we now use RKE `v1.6.0-rc4`?

File: go.mod
Comment (kinarashah) [inline go.mod]: Could we use latest `v1.9.0-rc.1` tags for *-operators?

File: go.mod
Comment (kinarashah) [inline go.mod]: Do we still need this fork?

**Review by kinarashah — CHANGES_REQUESTED** [review]
@krunalhinguu  We also need to update `kubeVersion` in Chart.yaml to add support for 1.29 https://github.com/rancher/rancher/blob/3c48a195381ad55f9fef7e9bf32b1593d43368ca/chart/Chart.yaml#L6

File: go.mod
Comment (krunalhinguu) [reply go.mod]: @kinarashah please check this commit, it being handled there

https://github.com/rancher/rancher/commit/373a1d6d7bd4b8a00f39a419b4f299a8a0a3f28b

File: go.mod
Comment (jiaqiluo) [inline go.mod:5]: This will be switched back to rancher/shepherd, right?

File: pkg/api/norman/customization/project/project_actions.go
Comment (jiaqiluo) [inline pkg/api/norman/customization/project/project_actions.go:50]: Instead of dropping the case and returning the default "unrecognized action" error,  could we return an error stating that the feature is deprecated? I think it will help eliminate some confusion on both our and users' side.

File: pkg/api/steve/catalog/types/rest.go
Comment (jiaqiluo) [inline pkg/api/steve/catalog/types/rest.go:69]: It looks like this struct is used internally, so it should be ok to change.

File: tests/validation/tests/v3_api/resource/terraform/k3s/master/install_k3s_master.sh
Comment (jiaqiluo) [inline tests/validation/tests/v3_api/resource/terraform/k3s/master/install_k3s_master.sh:48]: Question: could you explain why?

**Review by jiaqiluo — CHANGES_REQUESTED** [review]
Such a huge PR, good job! 
It looks good to me overall, and I have a few suggestions and questions.

File: go.mod
Comment (krunalhinguu) [reply go.mod:5]: Yup, once this pull request is merged, automation team will proceed to update rancher/shepherd, which has a dependency on rancher/apis. Afterward, I'll ensure that this is updated with the appropriate rancher/shepherd tag.

File: pkg/api/norman/customization/project/project_actions.go
Comment (krunalhinguu) [reply pkg/api/norman/customization/project/project_actions.go:50]: `PodSecurityPolicy` has been deprecated since version 1.25 and has now been completely removed in version 1.29. To prevent any confusion in future code, we are eliminating all references to `PodSecurityPolicy`. Retaining this portion of code could potentially lead to confusion down the line. Therefore, I believe it's best to completely remove it from the codebase.

File: tests/validation/tests/v3_api/resource/terraform/k3s/master/install_k3s_master.sh
Comment (krunalhinguu) [reply tests/validation/tests/v3_api/resource/terraform/k3s/master/install_k3s_master.sh:48]: I've reverted this change because there was some confusion, with some mistaking it for a Go version.

Thanks for highlighting this.

File: go.mod
Comment (krunalhinguu) [reply go.mod:5]: Just to keep track of this : https://github.com/rancher/rancher/issues/45564

Comment (jiaqiluo) [reply go.mod:5]: Cool, it sounds good to me.

File: pkg/api/norman/customization/project/project_actions.go
Comment (jiaqiluo) [reply pkg/api/norman/customization/project/project_actions.go:50]: I realize that a dedicated error message doesn't help a lot, so your solution is good.

**Review by jiaqiluo — APPROVED** [review]

**Review by kinarashah — APPROVED** [review]

**Comment by kinarashah** [comment]
Need @maxsokolovsky's review around PSP removal before merging.

File: pkg/controllers/managementuser/rbac/podsecuritypolicy/version.go
Comment (krunalhinguu) [reply pkg/controllers/managementuser/rbac/podsecuritypolicy/version.go:1]: should i mark this as resolve then?

Comment (snasovich) [reply pkg/controllers/managementuser/rbac/podsecuritypolicy/version.go:1]: Resolved per comment by @maxsokolovsky

**Comment by snasovich** [comment]
@krunalhinguu , thank you for updating the PR. Unfortunately CI is failing after latest force-push that updated error handling packages usage - https://drone-pr.rancher.io/rancher/rancher/39365/6/2. Could you please resolve this?

**Comment by krunalhinguu** [comment]
@snasovich looks like all the builds are failing in release-v2.9, I have to check the reason

**Review by maxsokolovsky — COMMENTED** [review]
A couple of questions:

1. Are there any ClusterRoles and ClusterRoleBindings with permissions to perform actions on PSPs that may need to be removed?
2. If yes, do we know if these bindings only act on PSP objects and no others?

I am not seeing any cleanup of these remaining objects, and we might want that.

Cleanup is usually done by running custom code or Job when the agent starts up. See this example: https://github.com/rancher/rancher/pull/42325

**Review by maxsokolovsky — COMMENTED** [review]
Another question is about CRDs, CRs, and PSP resources themselves. 

In case of CRDs, any deletions will trigger K8s to delete custom resources of that kind. So those PSP Bindings will get deleted when the CRD for them is deleted. But what about PSPs themselves? Those are not strictly CRDs but resources of a K8s group. Are they treated the same as CRs when a CRD is deleted? If the K8s API server doesn't know about PSPs, does it delete them all?

**Comment by snasovich** [comment]
@maxsokolovsky , thank you for the feedback.
I'll let @krunalhinguu respond materially but if there are non-trivial changes that need to be done (like creating migration that cleans up CRs/CRBs and any other objects referencing PSPs) we may want to split it into a separate issue to unblock progress and testing on base 1.29 support and unblock work on 1.30 support. That issue will still target the same `v2.9-Next1` milestone so it won't get missed. Having a separate issue will also ensure we have more specific test cases for these scenarios.

@krunalhinguu (cc: @mitulshah-suse ), please let us know if in fact we require these non-trivial changes to address @maxsokolovsky's concerns.

**Comment by mitulshah-suse** [comment]
Have created a tracking issue to check and cleanup PSP related resources if needed as part of the 1.29 upgrade.
https://github.com/rancher/rancher/issues/45621
It should not be a blocker for merging this PR though, so please confirm if we can go ahead and merge this PR.
@snasovich @maxsokolovsky @MbolotSuse

**Comment by snasovich** [comment]
@mitulshah-suse , yes, I'm on board with this approach per the reasoning given in the previous comment.
However:
1. @krunalhinguu , unfortunately we have another merge conflict since https://github.com/rancher/rancher/pull/45485 got merged. Could you please resolve it?
2. @maxsokolovsky , we do need formal approval of this PR from your team (again, considering https://github.com/rancher/rancher/issues/45621 will be addressed separately).

**Review by maxsokolovsky — APPROVED** [review]

**Pull request merged**
